main.go

```diff
@@ -140,6 +176,100 @@ func main() {
+		log.Fatalf("Failed to get instance of the helmrelease with %v", err)
+	} else if err != nil {
+		// Unexpectedly, the object that we are trying to delete was not found
+		// If this happens, we will log an error and return
+		log.Errorf("Did not find the helm release: %v", hr.Name)
+		cancel()
+		return
+	} else {
+		instance.Annotations = hr.Annotations
+		instance.Labels = hr.Labels
+		instance.Spec = hr.Spec
+
+		log.Infof("Found the helm release: %v, deleting the release...", hr.Name)
+		// Delete the HelmRelease
+		if err := retry(ctx, func() error { return clientSet.Delete(ctx, instance) }, interval); err != nil {
+			log.Fatalf("retry got err: %v", err)
+		}
+	}
+
+	pollingFunc := func() bool {
+		instance := &fluxhelmv2beta1.HelmRelease{}
+		key := client.ObjectKey{
+			Name:      hr.Name,
+			Namespace: hr.Namespace,
+		}
+		if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+			if ctx.Err() == context.DeadlineExceeded {
+				log.Errorf("Timed out polling for the helm release presence")
+			} else {
+				log.Fatalf("Failed to get instance of the helmrelease with %v", err)
+			}
+			return false
+		} else if err != nil {
+			log.Infof("Did not find the helm release: %v, the object is deleted", hr.Name)
+			return true
+		} else {
+			log.Infof("Found the helm release: %v, the object is not deleted", hr.Name)
+			return false
+		}
+	}
+
+	if err := poll(ctx, pollingFunc, time.Second); err != nil {
+		log.Errorf("Failed to poll the deletion of the helm release with %v", err)
+		// The helm release is still not deleted. Force cleanup as the final attempt
+		forceCleanupHelmRelease(clientSet, hr, pollingFunc, timeout)
+	}
+	cancel()
+}
+
+// remove the finalizer from the helm release and wait for it get garbage collected
+// if the helm release is still present, call 'helm uninstall' on it
+func forceCleanupHelmRelease(clientSet client.Client, hr *fluxhelmv2beta1.HelmRelease, poller func() bool, timeout time.Duration) {
+	// Create a new context with timeout
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	instance := &fluxhelmv2beta1.HelmRelease{}
+	key := client.ObjectKey{
+		Name:      hr.Name,
+		Namespace: hr.Namespace,
+	}
+	if err := clientSet.Get(ctx, key, instance); client.IgnoreNotFound(err) != nil {
+		log.Fatalf("Failed to get instance of the helmrelease with %v", err)
+	} else if err != nil {
+		// The helm release was not found after the first failed cleanup attempt
+		log.Errorf("Did not find the helm release: %v, the object is deleted", hr.Name)
+	} else {
+		log.Infof("Removing the helm release finalizer: %v...", fluxhelmv2beta1.HelmReleaseFinalizer)
+		controllerutil.RemoveFinalizer(instance, fluxhelmv2beta1.HelmReleaseFinalizer)
+		if err := clientSet.Update(ctx, instance); err != nil {
+			log.Fatalf("Failed to update the helmrelease with %v", err)
+		}
+		if err := poll(ctx, poller, time.Second); err != nil {
```

Comment:
Why are we polling again? Why not just execute the force uninstall and exit?

Reply:
I thought the design was to **wait** for the hr get garbage collected, but am ok with not wait. I think we are trying too much anyway :).

Reply:
Sure let's drop the poll

Reply:
Yes, already dropped the poll in the previous iteration.